protocols/gossipsub/src/handler.rs

```diff
@@ -250,10 +249,31 @@ impl EnabledHandler {
             ) {
                 // outbound idle state
                 Some(OutboundSubstreamState::WaitingOutput(substream)) => {
-                    if let Some(message) = self.send_queue.pop() {
-                        self.send_queue.shrink_to_fit();
-                        self.outbound_substream =
-                            Some(OutboundSubstreamState::PendingSend(substream, message));
+                    if let Poll::Ready(Some(mut message)) = self.send_queue.poll_next_unpin(cx) {
+                        match message {
+                            RpcOut::Publish {
+                                message: _,
+                                ref mut timeout,
+                            }
+                            | RpcOut::Forward {
+                                message: _,
+                                ref mut timeout,
+                            } => {
+                                if Pin::new(timeout).poll(cx).is_ready() {
+                                    // Inform the behaviour and end the poll.
+                                    self.outbound_substream =
+                                        Some(OutboundSubstreamState::WaitingOutput(substream));
+                                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                                        HandlerEvent::MessageDropped(message),
+                                    ));
+                                }
+                            }
+                            _ => {} // All other messages are not time-bound.
+                        }
```

Comment:
Would it make sense to move the logic for draining stale messages to the top? Then this check here wouldn't be needed anymore I think.